Add private preview API for Azure Image Builder

paulmey opened the pull request:
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

/CC @danielsollondon @johngossman

paulmey committed:
Add private preview API for Azure Image Builder

AutorestCI commented:
# Automation for azure-sdk-for-js

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-ruby

azuresdkci commented:
Can one of the admins verify this patch?

paulmey commented:
Spec has been reviewed previously by ARM team and service is in private preview.

AutorestCI commented:
# Automation for azure-sdk-for-java

ravbhatnagar commented on specification/imagebuilder/resource-manager/Microsoft.VirtualMachineImages/preview/2018-02-01-preview/imagebuilder.json:
```diff
@@ -0,0 +1,799 @@
+            "$ref": "#/parameters/ImageTemplateNameParameter"
+          }
+        ],
+        "responses": {
```
Model the 200 response as well. If resource was being updated, 200 OK can be returned instead of 201 created which gets returned for 1st PUT.

paulmey replied:
Does it matter that this never happens for this API version?

ravbhatnagar commented on specification/imagebuilder/resource-manager/Microsoft.VirtualMachineImages/preview/2018-02-01-preview/imagebuilder.json:
```diff
@@ -0,0 +1,799 @@
+            "$ref": "#/parameters/ImageTemplateNameParameter"
+          }
+        ],
+        "responses": {
```
"default" response which captures all the error responses is missing. The schema complies with the error contract defined in ARM RPC. Reference batch RP swagger.

ravbhatnagar commented on specification/imagebuilder/resource-manager/Microsoft.VirtualMachineImages/preview/2018-02-01-preview/imagebuilder.json:
```diff
@@ -0,0 +1,799 @@
+            "$ref": "#/parameters/ImageTemplateNameParameter"
+          }
+        ],
+        "responses": {
```
200 OK response. 204 should be returned when the request was well formed by the rsource was not found. 200 ok when it was successfully deleted.

ravbhatnagar commented on specification/imagebuilder/resource-manager/Microsoft.VirtualMachineImages/preview/2018-02-01-preview/imagebuilder.json:
```diff
@@ -0,0 +1,799 @@
+            "$ref": "#/parameters/ImageTemplateNameParameter"
+          }
+        ],
+        "responses": {
```
what is final response of this action? Model the 200 response.

paulmey replied:
This never returns a 200, it only starts an async operation.

ravbhatnagar commented on specification/imagebuilder/resource-manager/Microsoft.VirtualMachineImages/preview/2018-02-01-preview/imagebuilder.json:
```diff
@@ -0,0 +1,799 @@
+      }
+    }
+  },
+  "paths": {
```
APIs to list all imageTemplate under a resource group and under a subscription missing.

ravbhatnagar commented on specification/imagebuilder/resource-manager/Microsoft.VirtualMachineImages/preview/2018-02-01-preview/imagebuilder.json:
```diff
@@ -0,0 +1,799 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.VirtualMachineImages/locations/{location}/operations/{operationId}": {
```
dont need to model this in swagger, it comes from header.

sanjaiganesh replied:
@ravbhatnagar  can you please expand ? (for my understanding)

paulmey replied:
So this implies the response contract for the url in the header?
```
        "x-ms-long-running-operation": true,
        "x-ms-long-running-operation-options": {
          "final-state-via":"azure-async-operation"
        }
```
I'm good with that. :-)

ravbhatnagar commented on specification/imagebuilder/resource-manager/Microsoft.VirtualMachineImages/preview/2018-02-01-preview/imagebuilder.json:
```diff
@@ -0,0 +1,799 @@
+          "description": "The name to be used for the associated RunOutput.",
+          "pattern": "^[A-Za-z0-9-_]{1,64}$"
+        },
+        "tags": {
```
why cant the top level "tags" property be used for this?

majastrz replied:
I think Gaurav's question was more about naming being confusing with the usual resource "tags" property. Is it possible to change the name?

---
In reply to: [252806409](https://github.com/Azure/azure-rest-api-specs/pull/5055#discussion_r252806409) [](ancestors = 252806409)

paulmey replied:
changed to 'artifactTags'

ravbhatnagar commented on specification/imagebuilder/resource-manager/Microsoft.VirtualMachineImages/preview/2018-02-01-preview/imagebuilder.json:
```diff
@@ -0,0 +1,799 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.VirtualMachineImages/imageTemplates/{imageTemplateName}/run": {
```
what will this action do? What artifacts are being created? Do they not get created in the PUT?

paulmey replied:
This 'executes' the image build pipeline and create artifacts (ARM and storage) for each 'distribute' element in the template definition. Only one of these operations can be ongoing at a time. The result is referenced from the runOutputs proxy resources.

ravbhatnagar reviewed (commented):
This was reviewed in person long back with John and ARM team. Added some comments and excuse if some of these were discussed earlier as well.

paulmey commented:
Thanks for the review @ravbhatnagar. I'll push some changes ASAP.
/cc @sumit-kalra @feiyushi @amydutta @garimakhulbe02

anuchandy reviewed (changes requested):
@paulmey please add readme file as discussed over email. Ref: https://github.com/Azure/azure-rest-api-specs/tree/master/specification/hanaonazure/resource-manager

paulmey committed:
Review comments

 - Added list-by-resource-group and list-by-subscription
 - Added default error response
 - Removed explicit operation url (implied by async contract)
 - Adjusted 'delete' response codes to match expectations
 - Allow patch for tags
 - Added provider operations API

majastrz commented on specification/imagebuilder/resource-manager/Microsoft.VirtualMachineImages/preview/2018-02-01-preview/imagebuilder.json:
```diff
@@ -0,0 +1,999 @@
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
+        }
```
You should have this on all GET collection responses even if you're not planning to implement paging currently. It will allow SDKs to be forward-compatible if you ever change your mind about paging.

paulmey replied:
Changed paging to use nextLink on all collections/lists.

paulmey committed:
Review comment: use "nextLink" for paging

paulmey committed:
Adding readme files for SDK generation

paulmey committed:
PR: rename 'tags' -> 'artifactTags' for distributor

paulmey commented:
@anuchandy, I think I addressed your requested changes, PTAL. Thanks!

paulmey committed:
Add examples, fix minor issues in swagger

paulmey commented:
Ping

anuchandy commented:
@paulmey LGTM
